### PR TITLE
add tool to create firmware image header

### PR DESCRIPTION
**Pull request by vvfedorenko**
Driver has special check to avoid flashing wrong image, but there
was no tooling to create proper header for firmware image. This
simple tool can add required information.

### CODE REVIEW

File: Time-Card/TOOL/main.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	c := &Config{}
+
+	var logLevel string
+	flag.BoolVar(&c.Apply, "apply", false, "Create new firmware file with the header in the beginning")
+	flag.StringVar(&c.InputPath, "input", "", "Path to raw firmware file")
+	flag.StringVar(&c.OutputPath, "output", "", "Path to firmware file with header (will be overwritten")
+	flag.IntVar(&c.VendorId, "vendor", 0, "PCI VEN_ID to add to header")
+	flag.IntVar(&c.DeviceId, "device", 0, "PCI DEV_ID to add to header")
+	flag.IntVar(&c.HardwareRevision, "hw", 0, "PCI REV_ID (Hardware revision) to add to header")
+	flag.StringVar(&logLevel, "loglevel", "info", "Set a log level. Can be: debug, info, warning, error")
```
Comment (leoleovich) [inline Time-Card/TOOL/main.go:42]: You don't have any other log level used. Do we really need this?

File: Time-Card/TOOL/main.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+		if err != nil {
+			log.Fatal(err)
+		}
+		h := new(bytes.Buffer)
+		binary.Write(h, binary.BigEndian, &hdr)
+		_, err = outputFile.Write(h.Bytes())
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
```
Comment (leoleovich) [inline Time-Card/TOOL/main.go:110]: This should all go to a separate function and covered with some very basic tests.

File: Time-Card/TOOL/main.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	if c.Apply {
+		h := new(bytes.Buffer)
+		binary.Write(h, binary.BigEndian, &hdr)
+		_, err := outputFile.WriteAt(h.Bytes(), 0)
+		if err != nil {
+			log.Fatal(err)
+		}
+		outputFile.Close()
+		fmt.Printf("Firmware image with header is written to %s\n", c.OutputPath)
+	}
```
Comment (leoleovich) [inline Time-Card/TOOL/main.go:144]: This should go to another function (`crc or whatever) and have separate tests set.

File: Time-Card/TOOL/main.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+		log.Fatalf("Empty or incorrect PCI Vendor ID specified")
+	}
+
+	if c.DeviceId <= 0 || c.DeviceId > 65535 {
+		log.Fatalf("Empty or incorrect PCI Device ID specified")
+	}
+
+	if c.HardwareRevision < 0 || c.HardwareRevision > 65535 {
+		log.Fatalf("Incorrect PCI Device Revision ID specified")
+	}
```
Comment (leoleovich) [inline Time-Card/TOOL/main.go:76]: This should go to some validation function and be tested as well.

File: Time-Card/TOOL/main.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+type Header struct {
+	Magic            [4]byte
+	VendorId         uint16
+	DeviceId         uint16
+	ImageSize        uint32
+	HardwareRevision uint16
+	CRC              uint16
+}
+
+func main() {
```
Comment (leoleovich) [inline Time-Card/TOOL/main.go:32]: Ideally main should only have flag/library call and the rest should go to ./header/*.go

**Review by leoleovich — CHANGES_REQUESTED** [review]
Please keep in mind this is gonna be a very important tool which will have a large blast radius. It has to be tested extremely well before we provide it to other vendors.

File: Time-Card/TOOL/header/firmware.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	stat, err := c.InputFile.Stat()
+	if err != nil {
+		return 0, err
+	}
+
+	return uint32(stat.Size()), nil
+
+}
+
+func PrepareHeader(c *Config) (*Header, error) {
```
Comment (leoleovich) [inline Time-Card/TOOL/header/firmware.go:29]: <I am a linter>: All public methods should have a comment

File: Time-Card/TOOL/main.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	log.SetLevel(log.InfoLevel)
+
+	if err := h.CheckConfig(c); err != nil {
+		log.Fatal(err)
+	}
+
+	if err := h.OpenFiles(c); err != nil {
+		log.Fatal(err)
+	}
+
```
Comment (leoleovich) [inline Time-Card/TOOL/main.go:30]: You probably want `defer h.CloseFiles(c)` to make sure files always get closed?

File: Software/tft/header/config.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+	DeviceId         int
+	HardwareRevision int
+	InputFile        *os.File
+	OutputFile       *os.File
+}
+
+// Check config structure for valid values
+// PCI Vendor ID, Device ID should fit into uint16 and be positive
+// PCI Hardware Revision should be any uint16
+func CheckConfig(c *Config) error {
```
Comment (leoleovich) [inline Software/tft/header/config.go:22]: Correct way (to generate godoc automatically) is to start a comment with the method name:
> CheckConfig checks config structure...

File: Software/tft/main.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+	flag.BoolVar(&c.Apply, "apply", false, "Create new firmware file with the header in the beginning")
+	flag.StringVar(&c.InputPath, "input", "", "Path to raw firmware file")
+	flag.StringVar(&c.OutputPath, "output", "", "Path to firmware file with header (will be overwritten")
+	flag.IntVar(&c.VendorId, "vendor", 0, "PCI VEN_ID to add to header")
+	flag.IntVar(&c.DeviceId, "device", 0, "PCI DEV_ID to add to header")
+	flag.IntVar(&c.HardwareRevision, "hw", 0, "PCI REV_ID (Hardware revision) to add to header")
+	flag.Parse()
+
+	log.SetLevel(log.InfoLevel)
```
Comment (leoleovich) [inline Software/tft/main.go:21]: This is a default, no need to set.

File: Software/tft/header/config_test.go [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+		OutputPath:       "",
+		VendorId:         0,
+		DeviceId:         0,
+		HardwareRevision: -1,
+		InputFile:        nil,
+		OutputFile:       nil,
+	}
+
+	err := CheckConfig(c)
+	require.Error(t, err, "expect error on empty input filename")
```
Comment (leoleovich) [inline Software/tft/header/config_test.go:22]: Just FYI, the comment is optional

Comment (vvfedorenko) [reply Software/tft/header/config_test.go:22]: Yeah, I know, that was intended to provide our expectations for config

**Pull request merged**